src/mutual_information_sensitivity.jl

```diff
@@ -0,0 +1,272 @@
+The `MutualInformationResult` object contains the following fields:
+- `S1`: First order sensitivity indices.
+- `S1_Conf_Int`: Confidence intervals for the first order sensitivity indices.
+- `S2`: Second order sensitivity indices.
+- `S2_Conf_Int`: Confidence intervals for the second order sensitivity indices.
+- `ST`: Total order sensitivity indices.
+- `ST_Conf_Int`: Confidence intervals for the total order sensitivity indices.
+
+For fields that are not calculated, the corresponding field in the result will be an array of zeros.
+
+### Example
+
+```julia
+using GlobalSensitivity
+
+function ishi_batch(X)
+    A= 7
+    B= 0.1
+    @. sin(X[1,:]) + A*sin(X[2,:])^2+ B*X[3,:]^4 *sin(X[1,:])
+end
+function ishi(X)
+    A= 7
+    B= 0.1
+    sin(X[1]) + A*sin(X[2])^2+ B*X[3]^4 *sin(X[1])
+end
+
+lb = -ones(4)*π
+ub = ones(4)*π
+
+res1 = gsa(ishi,MutualInformation(),[[lb[i],ub[i]] for i in 1:4],samples=15000)
+res2 = gsa(ishi_batch,MutualInformation(),[[lb[i],ub[i]] for i in 1:4],samples=15000,batch=true)
+```
+
+### References
+[^1]: Lüdtke, N., Panzeri, S., Brown, M., Broomhead, D. S., Knowles, J., Montemurro, M. A., & Kell, D. B. (2007). Information-theoretic sensitivity analysis: a general method for credit assignment in complex networks. Journal of The Royal Society Interface, 5(19), 223–235.
+[^2]: Datseris, G., & Parlitz, U. (2022). Nonlinear Dynamics, Ch. 7, pg. 105-119.
+"""
+struct MutualInformation <: GSAMethod 
+    order::Vector{Int}
+    nboot::Int 
+    conf_level::Real
+    n_bin_configurations::Int
+    n_samples_per_configuration::Int
+end
+
+MutualInformation(; order = [0, 1], nboot = 1, conf_level = 0.95, n_bin_configurations = 800, n_samples_per_configuration = 100) = MutualInformation(order, nboot, conf_level, n_bin_configurations, n_samples_per_configuration)
+
+struct MutualInformationResult{T1, T2, T3, T4}
+    S1::T1
+    S1_Conf_Int::T2
+    S2::T3
+    S2_Conf_Int::T4
+    ST::T1
+    ST_Conf_Int::T2
+end
+
+function _total_order_ci(Xi, Y, entropy_Y, discretization_entropy; n_bootstraps = 100, conf_level = 0.95)
+    # perform permutations of Y and calculate total order
+    mi_values = zeros(n_bootstraps)
+    est = ValueHistogram(Int(round(sqrt(length(Y)))))
+    for i in 1:n_bootstraps
+        Y_perm = Y[randperm(length(Y))]
+        
+        conditional_entropy = entropy(est, StateSpaceSet(Xi, Y_perm)) - entropy(est, Xi)
+        mi_values[i] = (entropy_Y - conditional_entropy) / (entropy_Y - discretization_entropy)
+    end
+
+    α = 1 - conf_level
+
+    return quantile(mi_values, [α/2, 1 - α/2])
+end
+
+function _first_order_ci(Xi, Y; n_bootstraps = 100, conf_level = 0.95)
+
+    # perform permutations of Y and calculate mutual information
+    mi_values = zeros(n_bootstraps)
+    est = ValueHistogram(Int(round(sqrt(length(Y)))))
+    for i in 1:n_bootstraps
+        Y_perm = Y[randperm(length(Y))]
```

Comment:
Isn't it more efficient to first make `Y_perm = copy(Y)` and then within the loop `shuffle!(Y_perm)`? because in this way you allocate two vectors each time.

Reply:
Yes, true. I changed it